src/archive.rs

```diff
@@ -243,19 +246,77 @@ impl<'a> EntriesFields<'a> {
             return Err(other("archive header checksum mismatch"))
         }
 
+        let data_size = try!(header.data_size());
         let size = try!(header.size());
+        let sparse_chunks = if let Some(gnu) = header.as_gnu_mut() {
+            let mut chunks = Vec::new();
+            for item in &gnu.sparse[..] {
+                if item.offset[0] != 0 && item.numbytes[0] != 0 {
+                    let off = try!(octal_from(&item.offset[..]));
+                    let nbytes = try!(octal_from(&item.numbytes[..]));
+                    if nbytes > 0 {
+                        chunks.push((off, nbytes));
+                    }
+                }
+            }
+            if gnu.isextended[0] == 1 {
+                loop {
+                    let mut ext_header = [0u8; 512];
+                    try!(read_all(&mut &self.archive.inner, &mut ext_header));
+                    self.next += 512;
+                    let sparse_header: &GnuExtSparseHeader;
+                    sparse_header = unsafe { transmute(&ext_header) };
+                    for item in &sparse_header.sparse {
+                        if item.offset[0] != 0 && item.numbytes[0] != 0 {
+                            let off = try!(octal_from(&item.offset[..]));
+                            let nbytes = try!(octal_from(&item.numbytes[..]));
+                            if nbytes > 0 {
+                                chunks.push((off, nbytes));
+                            }
+                        }
+                    }
+                    if sparse_header.isextended[0] == 0 {
+                        break;
+                    }
+                }
+            }
+            if chunks.len() > 0 {
+                // Ensure chunks are in order and sized properly
+                let mut offset = 0;
+                let mut bytes = 0;
+                for chunk in &chunks {
+                    if chunk.0 < offset {
+                        return Err(other(
+                            "out of order or overlapping sparse chunks"))
+                    }
+                    bytes += chunk.1;
+                    offset = chunk.0 + chunk.1;
+                }
+                if bytes != data_size {
+                    return Err(other("wrong total length of data for sparse file"))
+                }
+                if offset > size {
+                    return Err(other("sparse chunk end exceeds file size"))
+                }
```

Comment:
It might be best to perform this validation as we're reading the chunks above as well, may help some ddos like situations as well perhaps